### ISSUE TITLE
mixin parameters affect variable of out scope

### ISSUE THREAD

**Issue by hax** [pug-php/pug#47]
It's a very serious bug...

``` pug

mixin test(a)
    div

- var_dump(a)
+test(a[0])
- var_dump(a)
+test(a[1])
- var_dump(a)

```

send `'a' => [1, 2, 3]`, you get:

```
array (size=3)
  0 => int 1
  1 => int 2
  2 => int 3
int 1
null
```

😱

**Comment by kylekatarnls** [comment]
Thanks, it seems this regression occurs since I implement the scope propagation that allow to use outer variable inside mixins (like in JavaScript and so in Pug.js) But indeed, we should exlude arguments from thoses varaibles.

Before we fix it, you can disable scope propagation with restrictedScope option:

``` php
$pug = new Pug\Pug(array(
    'restrictedScope' => true,
));
```

**Comment by hax** [comment]
@kylekatarnls Thank you for your response. We are the heavy user of original jade-php and we maintained a private forked version of it. I'm now investigating the possibility to migrate to pug-php. I found pug-php have changed a lot from original jade-php, for example use `get_defined_vars` and `extract` to pass the scope, though it seems can help to drop the old `attributes` hack, I'm worry about the performance of this method. Do you have any benchmark of that?

**Comment by kylekatarnls** [comment]
This will depend on the variables size. If you can keep small-sized variables, you have nothing to worry about, else, just use `'restrictedScope' => true,` to get the old behaviour. If you do not use "global" variables in your mixins, this option is recommended.

Currently the pug-php performance is very close to some raw PHP templates equivalents: 0,062 milliseconds per line of Pug code once the file is already cached (this is always the case if you use `->cacheDirectory()` on deployment).

It means, if you have a 300-lines pug template extending a 100-lines layout and including 2 x 100-lines pug files, it will takes 37 milliseconds to render.

I use pug-php in many sites and get no performance issues.

But I try my best to make cached files smaller and faster, it will be improved again in next versions.

**Comment by kylekatarnls** [comment]
About your project, I hope pug-php will match your needs. The issues you opened are much appreciated. I think pug-php can benefit your feedback as you can benefit the advantages of an open-source solution.

**Comment by kylekatarnls** [comment]
Fixed in master branch, soon released with 2.2

**Comment by kylekatarnls** [comment]
2.2 released, please update.
